include/torch-mlir/Dialect/Torch/IR/TorchOps.td

```diff
@@ -376,9 +376,6 @@ def Torch_PrimListUnpackOp: Torch_Op<"prim.ListUnpack", [
 
 def Torch_PrimTupleConstructOp: Torch_Op<"prim.TupleConstruct", [
     Pure,
-    TypesMatchWith<"contained types correspond to operand types",
-    "elements", "result", "Torch::TupleType::get($_ctxt, llvm::to_vector<6>($_self))",
-    "isValidSubtype">,
```

Comment:
I'm not sure what the right replacement is for this trait. With this trait, this is the build error:
```
/staging/torch-mlir/build/tools/torch-mlir/include/torch-mlir/Dialect/Torch/IR/TorchOps.cpp.inc:79294:66: error: no matching function for call to 'to_vector'
  ::mlir::Type odsInferredType0 = Torch::TupleType::get(context, llvm::to_vector<6>(operands[0].getType()));
                                                                 ^~~~~~~~~~~~~~~~~~
/staging/torch-mlir/externals/llvm-project/llvm/include/llvm/ADT/SmallVector.h:1298:46: note: candidate template ignored: substitution failure [with Size = 6, R = mlir::Type]: no matching function for call to 'begin'
SmallVector<ValueTypeFromRangeType<R>, Size> to_vector(R &&Range) {
                                             ^
/staging/torch-mlir/externals/llvm-project/llvm/include/llvm/ADT/SmallVector.h:1302:40: note: candidate template ignored: invalid explicitly-specified argument for template parameter 'R'
SmallVector<ValueTypeFromRangeType<R>> to_vector(R &&Range) {
                                       ^
```

Reply:
Hm, the use of `to_vector` here is a bit confusing to me, since `$_self` is a single type, not a range of types. 

What if you remove the `llvm::to_vector<6>` part and just do `{$_self}`? `TupleType::get` takes an `ArrayRef` as the second argument, so I'm assuming the list will get converted to the right type. Alternatively you can also use the `ArrayRef` constructor: `ArrayRef<...>{$_self}`

Reply:
No luck. Fixing the problem on the currently indicated line (either solution), breaks a previous line in verifyInvariantsImpl(). That is using getODSOperands(0), which returns a ::mlir::Operation::operand_range which has no cast to ArrayRef<::mlir::Type>.

 verifyInvariantsImpl has $_self = getODSOperands(0).getType() and inferReturnTypes uses $_self = operands[0].getType().

```
/staging/torch-mlir/build/tools/torch-mlir/include/torch-mlir/Dialect/Torch/IR/TorchOps.cpp.inc:79282:83: error: no matching conversion for functional-style cast from 'mlir::OperandRange::type_range' (aka 'ValueTypeRange<mlir::OperandRange>') to 'ArrayRef< ::mlir::Type>'
  if (!((isValidSubtype(Torch::TupleType::get(this->getOperation()->getContext(), ArrayRef<::mlir::Type>(this->getODSOperands(0).getType())), (*this->getODSResults(0).begin()).getType()))))
                                                                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Reply:
After talking to @jpienaar, he mentioned that a solution for now could be to implement the infer type interface for the `PrimTupleConstructOp`, which would prevent the autogenerated version that is causing the error. The long term solution is to add support for variadic operand + non-variadic return type in `TypesMatchedWith` upstream

Reply:
I tried adding the infer type interface to the op: `DeclareOpInterfaceMethods<InferTypeOpInterface>`, but that doesn't seem to have any effect on the creation of the inferReturnType method. I'll try to find why this is still happening and possibly we can discuss at the meeting on Monday.
I agree on the long term solution.

Reply:
elements is `Variadic<AnyTorchType>:$elements`, so I'd expect $_self to be a range here, but it actually becomes operands[0] instead of getODSOperands(0) with the former a value and the latter a range. Do you need anything other than the verification for this? If not, I'd just move this to a custom verify for the purposes of the integrate and follow up upstream.

Reply:
Yes, I think this is the right answer. I wasn't sure what the trait was giving us, but it appears to only be used for verification based on the tablegened files.